common/context/src/main/java/io/helidon/common/context/Contexts.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.common.context;
+
+import java.util.Optional;
+import java.util.Stack;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * Support for handling {@link io.helidon.common.context.Context} across thread boundaries.
+ */
+public final class Contexts {
+    static final ThreadLocal<Stack<Context>> REGISTRY = ThreadLocal.withInitial(Stack::new);
+
+    private Contexts() {
+    }
+
+    static void clear() {
+        REGISTRY.get().clear();
+    }
+
+    static void push(Context context) {
+        REGISTRY.get().push(context);
+    }
+
+    static Context pop() {
+        return REGISTRY.get().pop();
+    }
+
+    /**
+     * Get context registry associated with current thread.
+     *
+     * @return context that is associated with current thread or an empty context if none is
+     */
+    public static Optional<Context> context() {
+        Stack<Context> contextStack = REGISTRY.get();
+
+        if (contextStack.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.ofNullable(contextStack.peek());
+    }
+
+    /**
+     * Wrap an executor service to correctly propagate context to its threads.
+     *
+     * @param toWrap executor service
+     * @return a new executor service wrapping the provided one
+     */
+    public static ExecutorService wrap(ExecutorService toWrap) {
+        if (toWrap instanceof ContextAwareExecutorImpl) {
+            return toWrap;
+        }
+        if (toWrap instanceof ContextAwareScheduledExecutorImpl) {
```

Comment:
Is this necessary? `ContextAwareScheduledExecutorImpl` extends `ContextAwareExecutorImpl`, so it will be handled by the first `if` statement above -- it is effectively dead code, isn't it?

Reply:
Looks like it to me too, but I haven't had my coffee yet.

Reply:
Yep, comes from the time the `extends` was missing.